packer/images.json.pkr.hcl

```diff
@@ -0,0 +1,128 @@
+  instance_type = "t3.small"
+  region       = var.aws_region
+  secret_key   = var.aws_secret_key
+  source_ami   = data.amazon-ami.cos_ami.id
+  ssh_password = "cos"
+  ssh_username = "root"
+  launch_block_device_mappings {
+    device_name = "/dev/sda1"
+    volume_size = 15
+  }
+  user_data_file = "aws/setup-disk.yaml"
+  tags = {
+    Name    = "cOS"
+    Version = var.cos_version
+    Base_AMI_ID = "{{ .SourceAMI }}"
+    Base_AMI_Name = "{{ .SourceAMIName }}"
+  }
+}
+
+source "qemu" "cos" {
+  accelerator            = "${var.accelerator}"
+  boot_wait              = "${var.sleep}"
+  cpus                   = "${var.cpus}"
+  disk_interface         = "ide"
+  disk_size              = "${var.disk_size}"
+  format                 = "qcow2"
+  headless               = true
+  iso_checksum           = "none"
+  iso_url                = "${var.iso}"
+  qemuargs               = [["-m", "${var.memory}M"]]
+  shutdown_command       = "shutdown -hP now"
+  ssh_handshake_attempts = "20"
+  ssh_password           = "${var.root_password}"
+  ssh_timeout            = "5m"
+  ssh_username           = "${var.root_username}"
+  vm_name                = "cOS"
+}
+
+source "virtualbox-iso" "cos" {
+  boot_wait              = "${var.sleep}"
+  cpus                   = "${var.cpus}"
+  disk_size              = "${var.disk_size}"
+  format                 = "ova"
+  guest_additions_mode   = "disable"
+  guest_os_type          = "cOS"
+  headless               = true
+  iso_checksum           = "none"
+  iso_url                = "${var.iso}"
+  memory                 = "${var.memory}"
+  shutdown_command       = "shutdown -hP now"
+  ssh_handshake_attempts = "20"
+  ssh_password           = "${var.root_password}"
+  ssh_timeout            = "5m"
+  ssh_username           = "${var.root_username}"
+  vm_name                = "cOS"
+}
+
+# a build block invokes sources and runs provisioning steps on them. The
+# documentation for build blocks can be found here:
+# https://www.packer.io/docs/templates/hcl_templates/blocks/build
+build {
+  description = "cOS"
+
+  sources = ["source.amazon-ebs.cos", "source.qemu.cos", "source.virtualbox-iso.cos"]
+
+  provisioner "file" {
+    except = ["amazon-ebs.cos"]
+    destination = "/90_custom.yaml"
+    source      = "config.yaml"
+  }
+
+  provisioner "file" {
+    except = ["amazon-ebs.cos"]
+    destination = "/vagrant.yaml"
+    source      = "vagrant.yaml"
+  }
+
+  provisioner "shell" {
+    except = ["amazon-ebs.cos"]
+    inline = ["INTERACTIVE=false cos-installer --config /90_custom.yaml /dev/sda",
+      "if [ -n \"${var.feature}\" ]; then mount /dev/disk/by-label/COS_OEM /oem; cos-feature enable ${var.feature}; fi"
+    ]
+    pause_after = "30s"
+  }
+}
+
+  provisioner "shell" {
+    only = ["amazon-ebs.cos"]
+    inline = [
+      "cos-deploy --docker-image quay.io/costoolkit/releases-opensuse:cos-system-${var.cos_version}",
```

Comment:
I would just leave this open as an entire variable, like `container_image`. In this way the template can be used to bootstrap also images from external images too

Reply:
ah good point!

Reply:
like this? or the full url?

Thing is, we can take advantage of the cos version for several places

Reply:
I know, but this ideally should work also with custom images, not only for cOS ones, and we can't assume they are using `luet` for the building process. I would really let open to a cos-deploy-args even for now

Reply:
how about now?

Still, I think we are confusing things here. We are providing packer templates for our own use, to publish our images. We dont have an obligation to provide packer templates for derivatives or other customers.

We can provide examples, but those examples are not gonna be anything out of the ordinary from any example packer templates as they will use our published image as base, which is ready for consumption and then they can expand on that to build their own images following the packer documentation on how to provision stuff.

We need a clear separation here, because making OUR own stuff too flexible just in case someone uses it, only causes us identity issues. Is this for us, is this for derivatives? We should have a clear line here, this is for OUR deployment of images, here are also some examples on how to consume our images with packer.

Reply:
I thought that was clear from the card, but apparently it wasn't :sweat_smile: sorry ! 

The packer templates for Cloud are for derivatives/users only, as long as we don't need them internally to __provide__ base images. The fact that we use packer to spin up images used for tests is for convenience, but it is flexible enough to be re-used for derivatives too in case of need.

We could also use raw disks for tests for example, but then we don't have a way to see if we break the installation process.

Reply:
but...wont we need to publish images to cloud providers so others can consume them? ITs a different process altogether I think. In aws, are we gonna provide derivatives a recovery image instead of a fully released image??? Doesn't seem good to me, they will need to actually install the system and then once the image is "working" deploy their stuff.. I havent seen an image that works like that in a cloud provider, they are all ready images so your first launch with user data can configure it completely.

Reply:
what we are going to publish is only vanilla images where only the recovery partition is working. 

cOS is not meant to actually be used by per-se, but it's main use is to allow a complete system customization via container images. The recovery image just serves the purpose to bootstrap the minimal required to run the container image you supply (and taking later upgrades from)

Reply:
Ok, that clarifies it a lot! Then I think down the line we may need a refactor of the templates to separate everything so it's easier for derivatives to work with! Even a readme to provide a pck.vars file by the user with the proper vars instead of cli args and some vars descriptions